### DISCUSSION TITLE
Need easy way to update components

### DISCUSSION THREAD

**Discussion by pdeva** [shadcn-ui/ui#790] [Ideas]
Currently when I want to upgrade my components, i have to manually make a list of all my existing components and `add` command one by one for each one.

ideally the components themselves should be part of `components.json` file and a single command like `update` would update them to the latest version. i also think `theme` should be added to the config file too to allow easy experimentation with swtiching themes.

**Comment by beorn** [comment]
This adds/installs (and overwrites if existing) all shadcn components: `npx shadcn-ui@latest add -a -y -o`

**Reply by bryanlundberg** [reply]
Newest versions renamed the command to: `npx shadcn@latest add -a -y -o`

**Comment by remi2257** [comment]
You can use the following command line to update only installed components !
`for file in components/ui/*.tsx; do npx shadcn-ui@latest add -y -o $(basename "$file" .tsx); done`

Note: it can conflict a bit with components like Toast (one installation for multiple components)

**Reply by DenserMeerkat** [reply]
_PowerShell version of the same command_

`Get-ChildItem -Path components/ui/ -Filter *.tsx | ForEach-Object {
    $baseName = [System.IO.Path]::GetFileNameWithoutExtension($_.Name);
    npx shadcn-ui@latest add -y -o $baseName;
}`

**Reply by javaniecampbell** [reply]
This works great just an update, just remember to add semi-colons in the foreach:

Those using without `src/` directory
```powershell
Get-ChildItem -Path components/ui/ -Filter *.tsx | ForEach-Object { $baseName = [System.IO.Path]::GetFileNameWithoutExtension($_.Name); npx shadcn-ui@latest add -y -o $baseName; }
```

For those using `src/` directory
```powershell
Get-ChildItem -Path src/components/ui/ -Filter *.tsx | ForEach-Object { $baseName = [System.IO.Path]::GetFileNameWithoutExtension($_.Name); npx shadcn-ui@latest add -y -o $baseName; }
```
@DenserMeerkat  Thank you for the script, this is my thanks & contribution to that with gratitude.

**Reply by borgogelli** [reply]
> You can use the following command line to update only installed components ! `for file in components/ui/*.tsx; do npx shadcn-ui@latest add -y -o $(basename "$file" .tsx); done`
> 
> Note: it can conflict a bit with components like Toast (one installation for multiple components)

@remi2257 , please could you explain better what does "one installation for multiple components" means ?

**Reply by hughlv** [reply]
One note: This command should be run under project root, so if there is src/ folder, you should add 'src/' to the command:

```bash
for file in src/components/ui/*.tsx; do npx shadcn-ui@latest add -y -o $(basename "$file" .tsx); done
```

**Reply by feedm3** [reply]
with `pnpm` and the new `shadcn` package name:

```
for file in src/components/ui/*.tsx; do pnpm dlx shadcn@latest add -y -o $(basename "$file" .tsx); done
```

**Reply by pranta-barua007** [reply]
> You can use the following command line to update only installed components ! `for file in components/ui/*.tsx; do npx shadcn-ui@latest add -y -o $(basename "$file" .tsx); done`
> 
> Note: it can conflict a bit with components like Toast (one installation for multiple components)

is there any way to do so for custom registry?

**Comment by yahia-berashish** [comment]
I think that can be done by using the `shadcn-ui add` command then selecting the components you want to add @pdeva

**Comment by marcuslindfeldt** [comment]
The backstage project has an [upgrade helper](https://backstage.github.io/upgrade-helper/) that I think could be very useful for shadcn to see what has changed between versions. If it was integrated into the docs and on a component level it would be even better.

**Reply by Crystal-RainSlide** [reply]
Shadcn made a cli command called diff that work like that: https://ui.shadcn.com/docs/changelog#diff-experimental

But it's still experimental and is not working for me. Maybe it can't diff the small updates that happened in-between released verisons yet.

**Reply by marcuslindfeldt** [reply]
Seems nice but its not exactly the same I think. A lot of people tweak the shadcn components to match their brand and with this cli diff all of those changes will be highlighted as well. But I'm really only interested to see what changed between version a..b

**Reply by Crystal-RainSlide** [reply]
One more niche problem: shadcn/ui is using the latest main (instead of released versions) for installation. One need to have the exact git commit id of his installation to get a useful diff!  :(

---

To make things easier, I tagged every local tweak with comments, and I save the original code in those comments too. Thus my update process is just like a upstream merging.

**Comment by lianghx-319** [comment]
If we custom styled the built-in components from source code. Is any way to update to latest version and keep our own changed?

**Reply by MomentusCritisae** [reply]
I'm also very interested in this feature

**Reply by AndreaBarghigiani** [reply]
There's no way up until now to do that. Just tested with the `progress.tsx` component and if you ever change something the only option you have when you run `pnpm dlx shadcn@latest add progress` is to override it.

I'm kinda getting *"worried"* about this because I see many `React.ElementRef` deprecation warning around many components and IMO it'll get serious in the future when all components need to get updated.

I've tried to [open a discussion about this](https://github.com/shadcn-ui/ui/discussions/1907) a while ago, if you feel like participating just comment on that discussion.

**Reply by dwolrdcojp** [reply]
> There's no way up until now to do that. Just tested with the `progress.tsx` component and if you ever change something the only option you have when you run `pnpm dlx shadcn@latest add progress` is to override it.
> 
> I'm kinda getting _"worried"_ about this because I see many `React.ElementRef` deprecation warning around many components and IMO it'll get serious in the future when all components need to get updated.
> 
> I've tried to [open a discussion about this](https://github.com/shadcn-ui/ui/discussions/1907) a while ago, if you feel like participating just comment on that discussion.

https://github.com/reactjs/react-codemod#remove-forward-ref

**Reply by jmaister** [reply]
This is the big problem of copying the code of the componentes into your repository (like shadcn-ui) instead of importing the components from a library (others).

**Reply by raythurnvoid** [reply]
if you can have have a diff you can ask an LLM to apply the changed required.

You need the diff of you overrides then download the new shadcn stuff and ask Sonnet to re-apply your overrides.

**Comment by milon27** [comment]
this is really important feature

**Comment by realtebo** [comment]
So you must
- manually update shadcn 
- manually update all dependencies of shadcn
- manually override each and very component?


Please. provide an upgrade command !!!!!!!!!!!

**Comment by realtebo** [comment]
I just tried

```
 for file in src/components/ui/*.tsx; do npx shadcn@latest add -y -o $(basename "$file" .tsx); done
✔ Checking registry.
✔ Installing dependencies.
ℹ Skipped 1 files: (files might be identical, use --overwrite to overwrite)
  - src/components/ui/avatar.tsx
...
```

why the `-o` flag is ignored? isn't that the flag for `overwrite` ?

also... isn't a way to use components.json !?

**Comment by gvsakhil** [comment]
Its very difficult to update components one by one. Looking for a command to do this...

**Comment by ghost** [comment]
This collects all .tsx filenames (without extension) in src/components/ui/ and passes them as arguments to shadcn add.
```powershell
 bunx --bun shadcn@latest add $(Get-ChildItem -Path src/components/ui/ -Filter *.tsx | ForEach-Object { [System.IO.Path]::GetFileNameWithoutExtension($_.Name) } | ForEach-Object { $_ }) -y
```

**Comment by benyamynbrkyc** [comment]
is there any progress on this?

**Comment by ozeebee** [comment]
Hi, I'm late to the party, but isn't there already an update command ?
I just ran:
```shell
bun x shadcn-svelte@next update --all --yes
```
and it seems all my existing components were updated.
More info with `bun x shadcn-svelte@next update --help`

**Comment by m3h3d1ha2an** [comment]
<a href="https://x.com/shadcn/status/2000251574824882306?s=20"><img width="723" height="199" alt="screenshot-2025-12-15_20-26-40" src="https://github.com/user-attachments/assets/25fef1d7-da8e-4b6e-b35e-235770036bb7" /></a>

**Reply by fsolarv22** [reply]
This is the main reason why I don't use Shadcn; it's great that they're finally considering it. The way things are today, you need maximum agility in everything.

**Reply by ppulwey** [reply]
Tbh using BaseUI and let AI setting up some styled components based on it is not a big deal anymore. That's what I do for my customers projects now. It would still be nice to be able to use shadcn again in the future.

**Comment by ppulwey** [comment]
See my comment here: https://github.com/shadcn-ui/ui/issues/8359#issuecomment-3818359097

**Comment by AlejandroYYWireless** [comment]
Now that this is literally in the hands of Vercel, this should be done ASAP. This is insane we dont have this yet.

**Comment by MurzNN** [comment]
The critical issue with this feature is that we do not store anywhere the information about which exact versions of the Shadcn components were installed in the project, so I created a separate feature request to store this information: https://github.com/shadcn-ui/ui/discussions/10374

If we store this, we can just launch something like `shadcn diff` to see which exact components were changed locally by comparing the code with the remote files of the same versions (not like now with the latest versions) and see what exactly were changed and in what concrete components.  

So, if you have 100 components, but changed only 1 line in 1 component, this diff will show you only this single component and only this single changed line, not 100500 changes in other components, like it does now.

So, let's upvote the feature request https://github.com/shadcn-ui/ui/discussions/10374 to simplify the updating process!

**Comment by Prajwal-alphanimble** [comment]
you might wanna check out this discussion where I implemented a new resizable and dragable dashboard layout... this can be very helpfull to many developers in making analytic, sharable dashboard, the users can alter the dashboard based on their liking...

The disscussion link: https://github.com/shadcn-ui/ui/discussions/10466

**Comment by RobPruzan** [comment]
Git already solves all these problems fwiw. Someone could setup a mirror repo for just the components inside the shadcn repo. Then you could initialize a git sub module inside your components folder, and set the mirror repo as the remote. Then anytime there are new updates, you just pull. Anytime there's a conflict, you get a true merge conflict that is compatible with git tooling.
